### PR TITLE
Equalization refactoring

### PR DESCRIPTION
**Pull request by Paul-Licameli**
Resolves: *(direct link to the issue)*

Split up one of the largest single source files in Audacity into several smaller ones with non-cyclic dependencies.

This should make this complicated effect clearer and more maintainable.  It is a starting point before attempting the further problems of statelessness, real-time capability, and faster rendering with multiple cores.

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

**Review by vsverchinsky — APPROVED** [review]

File: src/effects/Equalization.cpp [hunk 1]
```diff
@@ -308,16 +287,6 @@ EffectEqualization::EffectEqualization(int Options)
    mWhenSliders[NUMBER_OF_BANDS] = 1.;
    mEQVals[NUMBER_OF_BANDS] = 0.;
 
-#ifdef EXPERIMENTAL_EQ_SSE_THREADED
-   bool useSSE;
-   GetPrivateConfig(CurrentSettingsGroup(), wxT("/SSE/GUI"), useSSE, false);
-   if(useSSE && !mEffectEqualization48x)
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply src/effects/Equalization.cpp:314]: All of that 48x stuff will be removed too.  It was an attempt at vectorizing the calculations someone wrote years ago but it doesn't compile on all machines and besides, I think there are better ways to exploit parallelism.

File: src/effects/Equalization.h [hunk 2]
```diff
@@ -90,12 +90,55 @@ class EQCurve
...
+};
+
 class EffectEqualization : public StatefulEffect,
                            public XMLTagHandler
 {
 public:
-   static inline EffectEqualization *
-   FetchParameters(EffectEqualization &e, EffectSettings &) { return &e; }
+   static inline auto
+   FetchParameters(EffectEqualization &e, EffectSettings &)
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply src/effects/Equalization.h:24]: See explanation in ShuttleAutomation.h.  It isn't a pure virtual function, but something like the compile-time equivalent -- an informal "concept."  Certain template expansions expect to pass that argument list.

Compare with other built-in effects.  Those that have been made stateless do not ignore the second parameter.  But these all now generate `FetchParameters` using the template `EffectWithSettings`

File: src/effects/Equalization.cpp [hunk 3]
```diff
@@ -559,6 +563,7 @@ std::unique_ptr<EffectUIValidator> EffectEqualization::PopulateOrExchange(
    const auto &mM = mParameters.mM;
    const auto &mLoFreq = mParameters.mLoFreq;
    const auto &mHiFreq = mParameters.mHiFreq;
+   const auto &mCurves = mCurvesList.mCurves;
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply src/effects/Equalization.cpp]: You know I believe strongly in preparing pull requests to be reviewed, understood, and convincingly free of mistakes, and that sometimes means overriding some style conventions if in return, I can made the diffs smaller and easier to review without a lot of extraneous changes.

Are you convinced that this sequence of commits achieves a splitting up of files but without any unintended changes of behavior?  That is what I want to know.

But if you like, I can restore more conventional variable names.  I do that now in a new commit.  Please review that too.

[the remark this answers is not in the document]
Comment (vsverchinsky) [reply src/effects/Equalization.cpp]: > Are you convinced that this sequence of commits achieves a splitting up of files but without any unintended changes of behavior? That is what I want to know.

Pretty much

**Review by vsverchinsky — APPROVED** [review]

File: src/effects/Equalization.cpp [hunk 4]
```diff
@@ -414,6 +414,31 @@ bool EffectEqualization::TransferDataToWindow(const EffectSettings &settings)
    return mUI.TransferDataToWindow(settings);
 }
 
+namespace {
+struct EqualizationTask {
```
Comment (vsverchinsky) [inline src/effects/Equalization.cpp:419]: I'd expect that entity called Task is a callable object or at least have `Run` or `Execute` or similar named method

File: src/PasteOverPreservingClips.h [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+class sampleCount;
+class WaveTrack;
+
+//! Substitute new contents into existing track, preserving clip boundaries
+/*!
+ @param start beginning position to paste over in oldTrack
+ @param len length to paste over in oldTrack
+ @param newContents begins at offset 0
+ */
+void PasteOverPreservingClips(
```
Comment (vsverchinsky) [inline src/PasteOverPreservingClips.h:25]: Isn't `WaveTrackUtils.h` suitable for utility functions like this one?

**Review by vsverchinsky — APPROVED** [review]

File: src/effects/Equalization.cpp (hunk 4 above)
Comment (Paul-Licameli) [reply src/effects/Equalization.cpp:419]: Wait for further developments... ;-)

File: src/PasteOverPreservingClips.h (hunk 5 above)
Comment (Paul-Licameli) [reply src/PasteOverPreservingClips.h:25]: True

**Pull request merged**